lib/class-wp-rest-shortcodes-controller.php

```diff
@@ -0,0 +1,189 @@
+<?php
+/**
+ * Shortcode Blocks REST API: WP_REST_Shortcodes_Controller class
+ *
+ * @package gutenberg
+ * @since 2.0.0
+ */
+
+/**
+ * Controller which provides a REST endpoint for Gutenberg to preview shortcode blocks.
+ *
+ * @since 2.0.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Shortcodes_Controller extends WP_REST_Controller {
+	/**
+	 * Constructs the controller.
+	 *
+	 * @since 2.0.0
+	 * @access public
+	 */
+	public function __construct() {
+		// @codingStandardsIgnoreLine - PHPCS mistakes $this->namespace for the namespace keyword
+		$this->namespace = 'gutenberg/v1';
+		$this->rest_base = 'shortcodes';
+	}
+
+	/**
+	 * Registers the necessary REST API routes.
+	 *
+	 * @since 0.10.0
+	 * @access public
+	 */
+	public function register_routes() {
+		// @codingStandardsIgnoreLine - PHPCS mistakes $this->namespace for the namespace keyword
+		$namespace = $this->namespace;
+
+		register_rest_route( $namespace, '/' . $this->rest_base, array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_shortcode_output' ),
+				'permission_callback' => array( $this, 'get_shortcode_output_permissions_check' ),
+			),
+			'schema' => array( $this, 'get_public_item_schema' ),
+		) );
+	}
+
+	/**
+	 * Checks if a given request has access to read shortcode blocks.
+	 *
+	 * @since 2.0.0
+	 * @access public
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_shortcode_output_permissions_check( $request ) {
+		if ( ! current_user_can( 'edit_posts' ) ) {
```

Comment:
If `$args['postId']` is not empty, then it should also check if `current_user_can( 'edit_post', $args['postId'] )`.